### PR TITLE
Add an initial visual identity check activity

### PR DESCRIPTION
**Pull request by ericcornelissen**
- Relevant Issues: #107
- Related Pull Requests: -

* * *

## What
A simple initial activity which can be used to verify the identity of another user while connecting via Bluetooth in a visual manner. For now the activity only contains a grid of buttons, later on the buttons will be overlayed with an image to present a more user friendly interface. Each button can be identified and after clicking on 5 buttons the activity will automatically go to the next activity (currently the LoginActivity) and provide it with the selected pattern.

## Why
As specified in #107 we need to create a secure way to pair with Bluetooth on top of the standard pairing method. This PR contains the first step towards creating this method.

## How
You can run the Cucumber tests and see it there or you can force the code to open the new activity by adding:
```java
Intent intent = new Intent(this, VisualVerificationActivity.class);
this.startActivity(intent);
```

## Alternative implementation
Not considered

## Notes
As mentioned before this PR only adds simple activity with some buttons. In future PRs there will be a few additions, for example:
- Have a 'waiting' activity for the slave while it is waiting for the master to input the pattern.
- Have a 'waiting' activity for the master while it is waiting for the slave to reproduce the pattern.
- Have an image on top of the buttons so the user can more easily specify what button it clicked (by specifying visually where they clicked).

### CODE REVIEW

**Comment by jverbraeken** [comment]
Nice work, Eric!

Have you already looked at another much simpler solution: drawing a single image and then using [touch events](http://stackoverflow.com/questions/3476779/how-to-get-the-touch-position-in-android) to get the location of where the user pressed. Such a solution is more elegant, less code and has a lot (!) more precision than 16 buttons (each having 2560x1440 / 16 = 230,400 times less precision than OnTouchEvent(x, y))

**Comment by ericcornelissen** [comment]
@jverbraeken I don't think we need that kind of precision to make this work, and doing it that way could complicate things significantly for no good reason.

The reason I don't think we need that kind of precision is because of the image that I image will be displayed on top of the buttons. This should be an image with a few distinct figures/drawings that can easily be communicated. These figures will be pretty large so it easy for user to identify them on the screen and touch them in order to replicate the pattern of the master device.

There is really only one downside to this approach, which is that all images have to adjusted for the buttons so that the distinct elements are within the borders of the underlying buttons.

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 1]
```diff
@@ -59,7 +59,7 @@ private void validateAccount() {
         if (skipPassword) {
             LOGGER.warn("Password skipped!");
             mError.setVisibility(View.GONE);
-            final Intent intent = new Intent(LoginActivity.this, MainActivity.class);
+            final Intent intent = new Intent(LoginActivity.this, VisualVerificationActivity.class);
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Could you change this back? It's better to only keep that for testing

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Sure, forgot about it 😅

File: app/src/androidTest/assets/features/visualVerification.feature [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  Do several things on the VisualVerificationActivity
+
+  Scenario Outline: Fill out a correct pattern
+    Given I am viewing the visual verification activity
+    When I press button <first_button>
+    And I press button <second_button>
+    And I press button <third_button>
+    And I press button <fourth_button>
+    And I press button <fifth_button>
+    Then I leave the visual verification activity
```
Comment (TheBonheurs) [inline app/src/androidTest/assets/features/visualVerification.feature]: Should you always leave this activity, even in the case your sequence wasn't the right one? Or is this for the person that initiates the connection?

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 3]
```diff
@@ -77,7 +77,12 @@ public void onClick(final View view) {
         }
 
         final ListView lv = (ListView) findViewById(R.id.listView);
-        lv.setAdapter(new ContactListAdapter(this, this.contacts));
+        try {
+            fillDatabaseWithDemoData();
+            lv.setAdapter(new ContactListAdapter(this, serviceLocator.getDatabase().getAllContacts()));
+        } catch (final IOException e) {
+            LOGGER.error("Failed to retrieve contacts from database", e);
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Filling the db could also go wrong,I don't mind this log but a more ambiguous one might be better.

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                    nextActivity();
+                } else {
+                    securityCode += button;
+                    LOGGER.info("code so far: " + securityCode);
+                }
+            }
+
+        };
+
+        this.findViewById(R.id.button00).setOnClickListener(onClickListener);
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: You could also use this inthe layout xml https://developer.android.com/reference/android/view/View.html#attr_android:onClick

File: app/src/androidTest/assets/features/visualVerification.feature (hunk 2 above)
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/visualVerification.feature]: For now this is intended for the user initiating the connection

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 3 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: I agree, but the `fillDatabaseWithDemoData();` should be removed in the future, thats why I exluded it from the log

Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Fine by me 👍

File: app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java (hunk 4 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/VisualVerificationActivity.java]: Good one 👍

**Pull request merged**